rfd/0061-tsh-aliases.md

```diff
@@ -0,0 +1,165 @@
+
+- Unlike `git` aliases explored above, proposal is to always require the binary
+  name instead of defaulting to `tsh` and requiring special syntax otherwise.
+- The `<command>` will be fed to `exec.Command` for execution. To chain multiple
+  commands, use pipes, etc. users will explicitly use `bash -c`.
+- The `<command>` will require `$@`, `$1`, etc. to reference the arguments
+  provided to the alias command.
+
+### Examples
+
+Starting with a very simple example, suppose you're tired of typing `tsh login`
+every morning and want to alias it to something shorter e.g. `tsh l`:
+
+```yaml
+aliases:
+    "l": "tsh login $@"
+```
+
+As a more practical example, the following alias will make `tsh login` default
+to the specific auth connector and username:
+
+```yaml
+aliases:
+    "login": "tsh login --auth=local --user=alice $@"
+```
+
+Similarly, as discussed in the search-based access requests RFD, some users
+may want `tsh ssh` to default to the mode in which it will auto-request access
+to the node upon encountering an access denied error:
+
+```yaml
+aliases:
+    "ssh": "tsh ssh -P $@"
+```
+
+Note: command arguments will be resolved prior to invoking the alias. So in the
+example above, when a user runs `tsh ssh root@node1`, the alias command executed
+will be `tsh ssh -P root@node1`.
+
+### Environment variables
+
+Each alias invocation will set a few environment variables that can be consumed
+within the alias.
+
+To prevent infinite recursion, `TSH_ALIAS` variable will be set indicating the
+command is invoked as an alias. If it's detected, `tsh` won't attempt to expand
+the same command again. For example, `TSH_ALIAS=login`. This will prevent other
+`login` commands from being expanded but will allow using aliases in other
+aliases (see below for an example).
+
+Each alias will also set `TSH` environment variable to the path of the `tsh`
+binary that invoked the alias.
+
+### More examples
+
+An alias can also define a custom subcommand that combines multiple commands.
+The following alias will connect to a node within a specific leaf cluster
+using `tsh connect leaf ubuntu@node-1` command:
+
+```yaml
+aliases:
+    "connect": "bash -c 'tsh login $1 && tsh ssh $2'"
+```
+
+The following alias will list nodes in all clusters:
+
+```yaml
+aliases:
+    "lsall": "bash -c 'for cluster in $(tsh clusters | tail -3 | head -2 | cut -d \' \' -f1); tsh ls --cluster=$cluster; done'"
+```
+
+Multiple aliases can be defined in the config and can reference each other:
+
+```yaml
+aliases:
+    "login": "$TSH login --auth=local --user=alice $@"
+    "ssh": "$TSH ssh -P $@"
+    "connect": "bash -c '$TSH login $1 && $TSH ssh $2'"
+```
+
+In this example, `tsh login` and `tsh ssh` will use the aliases when invoked
+as a part of the "connect" command.
+
+## Future work
+
+When we have `bash`/`zsh` auto-completion, it should take aliased commands into
+account as well. For reference, `git` autocompletion supports it.
```

Comment:
Would an option to have commands like `tsh-whatever` that are on the `$PATH`  become `tsh whatever` aliases be considered in the future?

git supports this and could be a bit simpler to distribute than config files

Reply:
I feel a lot of the use of the aliases will be very specific to people's use-cases e.g. wanting to tweak default behavior of a particular command. Otherwise, we'd probably just want to integrate the subcommands into `tsh` directly.